src/server.rs

```diff
@@ -96,6 +96,65 @@ where
     }
 }
 
+#[cfg(feature = "early-data")]
+impl<IO> TlsStream<IO>
+where
+    IO: AsyncRead + AsyncWrite + Unpin,
+{
+    pub fn poll_read_early_data(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        use std::io::Read;
+
+        let this = self.get_mut();
+        let mut stream =
+            Stream::new(&mut this.io, &mut this.session).set_eof(!this.state.readable());
+
+        match &this.state {
+            TlsState::Stream | TlsState::WriteShutdown => {
+                {
+                    let mut stream = stream.as_mut_pin();
+
+                    while !stream.eof && stream.session.wants_read() {
+                        match stream.read_io(cx) {
+                            Poll::Ready(Ok(0)) => {
+                                break;
+                            }
+                            Poll::Ready(Ok(_)) => (),
+                            Poll::Pending => {
+                                break;
+                            }
+                            Poll::Ready(Err(err)) => return Poll::Ready(Err(err)),
+                        }
+                    }
+                }
```

Comment:
This code looks it's duplicated from somewhere else. Can we abstract over it instead? If not, why not?